OO refactoring and git-blame preloading

uliska opened the pull request:
This has become more involved and intrusive than I'd prefer at this point of contribution to a repository, and I'm feeling somewhat uneasy about it. However, it became an all-or-nothing thing and wouldn't have made sense to send slice by slice.

This PR does more than one thing unforfunately:

* Replace the GitPython dependency with an object oriented abstraction with classes
    * GitCommand
    * Repo
    * Commit
    * Page
    * Author
* Preload the git-blame statistics in the `on_files` event

I think this restructuring will make it easier to add more functionality and is therefore useful.

It does *not* include a proposal for #9 yet although that should be trivial.

This commit completely breaks the unit tests, but I didn't want to put my hands on these before getting some feedback on the PR itself.

uliska commented on mkdocs_git_authors_plugin/plugin.py:
```diff
@@ -6,6 +6,22 @@ class GitAuthorsPlugin(BasePlugin):
+        contribution of an author to the repository can be
+        retrieved on *any* Markdown page.
+        """
+        cnt = 0
```
I'll remove the `cnt` at the next opportunity, they are left here accidentally

uliska commented:
I've now added th `{{ git_authors_list }}` feature proposed in #9.

```markdown
### Mitwirkende

Folgende Autoren haben an der Erstellung dieser Dokumentation mitgewirkt:

{{ git_authors_list }}
```
results in the following output:

![grafik](https://user-images.githubusercontent.com/1812148/75756950-c6e68c00-5d31-11ea-945f-8cae31c39987.png)

with the following configuration:

```yaml
- git-authors:
    show_contribution: true
    show_lines: true
    label_lines: Zeilen
    uncommitted_name: Nicht gesichert
```

The "Nicht gesichert" refers to just this example which hasn't been committed.

I think I would like to see an option `labels`, which allows more labels to be added in a more structured way.

*[This of course still breaks all unit tests.]*

uliska commented:
With the sort order I think I'm basically done with everything I'd like to have added to the plugin.

If you basically agree with the restructuring in this PR

* I'll take care of adding unit tests (actually I'd be glad if you could look after the existing ones, @timvink)
* I'll add the relevant documentation

I have a final suggestion for you, if you agree with my approach to removing GitPython support: I would merge this (as a PyPi package) with the `mkdocs-git-revision-date(-localized)` packages, but still expose them as separate MkDocs plugins. So all packages can share the basic `repo` module and its functionality.

uliska commented:
Sorry, yet another one ;-)

timvink commented:
Wow, I can see you definitely went for it ! Starting to enjoy the open source work @uliska ?

This is a *huge* PR, which also means I have a lot of comments. I focussed especially on maintainability and simplicity.  In general I think it's a great PR that will really help with future additions as well. I realise it addresses your specific needs, but it still needs more work to get to a place where it is very usably for a large group of people. 

I already mentioned them in the specific review comments, but main review points are:

- Reduce verbosity, some methods not used or necessary
- I am *much* more concerned about stability than adding an extra 850kb dependency on a very mature`GitPython`.  Let's seriously consider removing `GitCommand` and reintroduce `GitPython`.
- Separate concerns. Keep the git stuff in the git classes, and the (html) summary stuff in `util.py`
- I had to be a bit strict on some of the extra options and the extra markdown tag. Hope you understand. See motivation in the review comments.
- I have been more explicit on the 'roadmap' I envision for this plugin, see #16

> I have a final suggestion for you, if you agree with my approach to removing GitPython support: I would merge this (as a PyPi package) with the mkdocs-git-revision-date(-localized) packages, but still expose them as separate MkDocs plugins. So all packages can share the basic repo module and its functionality.

Thanks for the suggestion. I'd prefer to keep depending on `GitPython` for `mkdocs-git-revision-date-localized`. Copying the `repo.py` will mean more complex maintainability. Inevitably there are going to be specific bugs with new code.. and I'm already noticing some copy/paste troubles between the packages. You did some good work with the Git abstractions however. I would be open to depending on a separate new python package. There might be more demand by package authors to use a very simple and very lightweight Git wrapper (`GitPython` docs are too verbose and a bit unclear to be honest). Something for you to consider :)

Before working on adding docs and unit tests, let's address the review comments first OK ?

uliska commented:
> Starting to enjoy the open source work @uliska ?

Well, if you look at my [profile](https://github.com/uliska) you may realize that "starting" doesn't really pin it down ;-) (and that's only GitHub).

Rather I tend to become creative extending code with (generally) useful stuff - when there's something to build upon.

> I realise it addresses your specific needs, but it still needs more work to get to a place where it is very usably for a large group of people.

I'm aware of that. But things will only move on when started in the first place.

> I already mentioned them in the specific review comments, 

I seem to be missing something, but I don't see any review comments.

> Reduce verbosity, some methods not used or necessary

I assume I'll see what you mean once I see the review comments?

> Let's seriously consider removing GitCommand and reintroduce GitPython.

I suggest leaving that question to itself for a few days to see what comes out of contemplating it.

> You did some good work with the Git abstractions however. I would be open to depending on a separate new python package. There might be more demand by package authors to use a very simple and very lightweight Git wrapper (GitPython docs are too verbose and a bit unclear to be honest). Something for you to consider :)

I don't think converting my code into a standalone Git wrapper makes a lot of sense. The point of the exercise basically was to only implement what's actually needed for the "client" (the plugin). Once it is exposed to a "general audience" all kinds of requests would come, and it would end as a less-tested GitPython.

> Copying the repo.py will mean more complex maintainability. 

I will not push you on this, but maybe there's a misunderstanding here. I didn't talk about *copying* the code. I was thinking of *one* common Python package that includes *two* MkDocs plugins as exposed through different entry points in the `setup.py` file.

For today my time is completely taken up, but I'll consider your comments in more details when I get the chance to see your individual review comments. (maybe there's an option to finally publish a review?)

uliska commented on mkdocs_git_authors_plugin/repo.py:
```diff
@@ -0,0 +1,774 @@
+        entry['lines'] += lines
+        current_dt = entry.get('datetime')
+        commit_dt = commit.datetime()
+        if not current_dt or commit_dt > current_dt:
```
I think I have made things more confusing than necessary by naming two different things the same: `page`.

Usually you'd expect `self.page()` to return a `Page` object, either existing or newly created.
`Author.page()` is different and should therefore be renamed (although I'm not quite sure to what).

The `Author` object stores information about how many lines it changed on which pages. This is stored in a dictionary with the page paths as key. Each entry holds a Page object, the line count, and the datetime of the last modification as a datetime object and as a string representation. But initially no information on the datetime is stored in the dictionary.

This function is called while processing the Git blame information. An author "receives" a line from a page in a given commit.
First we look up the entry maintaining the page. 

When the Commit object is created there are no lines yet, therefore `current_dt` is initialized with `None`. This logic ensures that in the first iteration the datetime is set to the commit's timestamp while in each subsequent execution of the function the commit's timestamp is compared to the entry's.

BUT (and I've been wondering about this all along): If I'm not mistaken this information is not really being used anywhere. It ends up in the `author` dict in the page context. But I can't really figure out who would make use of the information when *a given author* last modified the current page.
If we could get rid of that information in the `author` dict this could be written much more concisely, there wouldn't even be the need for the `commit` argument.

timvink replied:
>But I can't really figure out who would make use of the information when a given author last modified the current page. If we could get rid of that information in the author dict this could be written much more concisely

I'm still thinking about how to merge this plugin with `mkdocs-revision-date-localized`. I researched if retrieving datetime information from authors in `git blame` would solve the problem where shallow clones (used in CI builds, see https://github.com/timvink/mkdocs-git-revision-date-localized-plugin/issues/10) make it hard to find last update date of a file. [It doesn't](http://git.661346.n2.nabble.com/Shallow-clones-vs-blame-log-td6082351.html). 

I don't see any other use case. 

So let's decide not to supply the page authors last modification date, and make things a bit simpler!

uliska replied:
> I don't see any other use case.

That had been my impression, actually.

But I suggest *not* to decide this right now.
*If* we decide to keep the Git classes (vs. GitPython) *and* to merge the two plugins at some point we may well retrieve the information in this function. While I see that it doesn't solve the issue of shallow copies (and upon first sight I have the impression there is no remedy for that problem) it could make yet another invocation of Git obsolete.

timvink replied:
I agree, let's leave it for now. We can always revisit / refactor with future new insights.

uliska commented on mkdocs_git_authors_plugin/repo.py:
```diff
@@ -0,0 +1,774 @@
+        Returns:
+            The full name as a string.
+        """
+        return self._name
```
I'd oppose that (as with the other comparable instances).

uliska commented on mkdocs_git_authors_plugin/plugin.py:
```diff
@@ -1,15 +1,81 @@
+from .repo import Repo
 
 class GitAuthorsPlugin(BasePlugin):
     config_scheme = (
```
Only after going through all the other comments I now see that this means that these have to be removed *now* so people won't start using them before the localization will be there to replace the functionality.

e6c6a8bb705fb9e02aae913ff1a27b9605b00777

uliska replied:
> I plan to add support for this plugin to mkdocs material (see squidfunk/mkdocs-material#1403).

Which would be great!

uliska commented:
I've gone through the review comments. Indeed Pull Requests should not be so heavy, but as said at one point I didn't see any chance for an in-between anymore.

I've marked many comments as "resolved", in most cases after addressing them with a commit. A number of comments are not immediately resolvable and I added comments and questions. One general remark: I've stated disagreement in quite a number of cases. Please don't get me wrong with these - I don't want to appear arrogant, and I don't want to fight over these issues. But it's important to clearly express differing opinions in order to be able to resolve them for the best. So I won't be offended if you disagree with my comments on your turn ...

There are two recurring issues where we either have different ideas about programming style or where you may not fully see my idea (which does not mean my idea is correct or the best solution).

A)
If I have a field
```
class MyClass:
    def __init__(self):
        self._field = 'XY'
```
I think the paradigm of encapsulation says it is always better to have a *method* `MyClass.field()` to access this than to directly access it through `my_class_object._field`. Therefore I would strongly suggest *not* to remove these "property" methods. Except in a few cases where they are not used at all, in which case I have removed them in new commits now.

B)
There are a number of methods that retrieve *or* create objects, which you seem to find somewhat unclear.
The more I thought about this I'm sure it's the right way to approach the task in the context of this plugin. It is a fairly common pattern for accessing complex and/or expensive objects. Create them upon first request and then store them.
I think it's also straightforward from a semantic perspective. I just take one example, but it's basically the same throughout: When processing the `git blame` data for a page I encounter a Git SHA for each line. To deal with it I need to fetch the `Commit` object corresponding to that SHA. So I'm asking for it, and if this SHA has been used already there's an object available, otherwise it will created (and the actual Git commit processed) at the time when I need the object for the first time.
I think this is not less clear than the more explicit way of asking for the object and if it's not there already create it in a separate step.

uliska commented on mkdocs_git_authors_plugin/repo.py:
```diff
@@ -0,0 +1,774 @@
+            Author object
+        """
+        if not self._authors.get(email, None):
+            self._authors[email] = Author(self, name, email)
```
After going through all this I'd be more favourable to my implementation (as explained in the "conversation" comment).

timvink replied:
I'm now also convinced your implementation is better. Nice for me to also see some new patterns!

timvink commented on mkdocs_git_authors_plugin/plugin.py:
```diff
@@ -66,13 +131,32 @@ def on_page_context(self, context, page, **kwargs):
+        """
+        Reference to the Repo object of the current project.
+        """
+        return self._repo
```
Like you mentioned, I think it is an idea of differing code styles. I know it is better not to direct access class attributes, but I prefer brevity more. Probably I have not worked on sufficiently complex code bases to see the added value. I won't push the issue though, it's minor. Let's leave this (and the other similar methods) as-is. I'll resolve the conversations.

timvink commented on mkdocs_git_authors_plugin/repo.py:
```diff
@@ -0,0 +1,774 @@
+        t = [int(v) for v in t.split(':')]
+        # timezone info looks like +hhmm or -hhmm
+        tz_hours = int(tz[:3])
+        th_minutes = int(tz[0] + tz[3:])
```
Makes sense. Unless you want to cleanly separate all git functionality, but that's not needed. Like you said earlier, these git classes are customized for use with mkdocs.

timvink commented on mkdocs_git_authors_plugin/repo.py:
```diff
@@ -0,0 +1,774 @@
+        show_lines = show_contribution and self.config('show_lines')
+        label_lines = self.config('label_lines')
+        result = """
+    <ul class='git-authors'>
```
No what I meant is spanning the entire git authors output. 

```html
<span class='git-authors'>
<!-- output of markdown tag -->
</span>
```

I learned this lesson here: https://github.com/timvink/mkdocs-git-revision-date-localized-plugin/issues/12 

mkdocs-material also `<span>`s and styles entire sections, so I thought it best to be consistent. For styling in css you can still style your lists with `.git-authors ul {}`

uliska replied:
OK, I see.

Is that the form you think it should have?
65d177999eac005b0dcc0445f16c7fabceeb15c4

timvink commented on mkdocs_git_authors_plugin/plugin.py:
```diff
@@ -1,15 +1,81 @@
+        This populates all the lines and total_lines properties
+        of the pages and the repository, so the total
+        contribution of an author to the repository can be
+        retrieved on *any* Markdown page.
```
I was referring to updating the docstring indeed. Next to the update you already now made, we can consider adding an extra line explaining that `on_files()` is called before the other events we use in this plugin, which means `self.repo` will contain already processed authors from the pages. That's the whole trick of this new setup, right?

uliska replied:
4a208dcafd0657e524ab30f1c5babf0f8c28216c

I don't mark this as resolved although the commit is probably sufficient.

uliska replied:
...
> That's the whole trick of this new setup, right?

Yes.

timvink commented on mkdocs_git_authors_plugin/repo.py:
```diff
@@ -0,0 +1,774 @@
+            m = line_pattern.match(line)
+            if m:
+                sha = m.group(1)
+                content = m.group(2).strip()
```
> I suggest that we complete polishing this PR and merge it.

Yes, I agree completely

> In the meantime I'll do two things

Great suggestions! I've also opened #17, which I can pick up after we complete this PR

uliska replied:
I've created https://stackoverflow.com/questions/60523415/robustly-retrieve-sha-and-line-content-with-git-blame-python3 but while finishing it and testing with `git blame --porcelain` I found that we *should* actually use that (assuming that porcelain output can be considered stable), for a specific reason:

The output of the porcelain version gives us all the information we also store in a `Commit` object. that means: by parsing the git-blame output I can pass all relevant information to `Commit`'s initializer, and we don't even need to run `git show` on the commit at all.

I'll give that a try.

timvink replied:
Good find!

I also found the [git blame implementation from GitPython](https://github.com/gitpython-developers/GitPython/blob/d7b401e0aa9dbb1a7543dde46064b24a5450db19/git/repo/base.py#L781) uses the porcelain output as well. Maybe you can re-use parts of their code for parsing it.

Perhaps it's better to create a separate issue + PR for this? Then we can resolve this discussion. I'd like to merge this PR to master. It will fail the build, but we won't release yet. Then I can work on the unit tests & other issues, building on this PR.

timvink replied:
Actually went ahead and created https://github.com/timvink/mkdocs-git-authors-plugin/issues/20. Resolving this discussion.

uliska replied:
I see. Unfortunately I can't assign issues to myself.

timvink commented on mkdocs_git_authors_plugin/repo.py:
```diff
@@ -0,0 +1,774 @@
+
+                if content or self.repo().config('count_empty_lines'):
+                    # assign the line to a commit and count it
+                    commit = self.repo().commit(sha)
```
Thanks for also explaining the rational where you both retrieve an ID and process/store it if not available yet. Interesting pattern.

timvink commented on mkdocs_git_authors_plugin/repo.py:
```diff
@@ -0,0 +1,774 @@
+        Returns:
+            Repo instance
+        """
+        return self._repo
```
And I only later realized `AbstractRepoObject` was used in multiple places. And we've already discussed accessing class attributes. Let's keep this :)

timvink commented on mkdocs_git_authors_plugin/repo.py:
```diff
@@ -0,0 +1,774 @@
+        Args:
+            number of lines to add, default: 1
+        """
+        self._total_lines += cnt
```
On second thought, I'd prefer to keep it as it is currently. Let's focus on finishing this PR, we can always improve this later (after discussing option 2 vs 3)

timvink commented on mkdocs_git_authors_plugin/repo.py:
```diff
@@ -0,0 +1,774 @@
+        """
+        self._total_lines += cnt
+
+    def author(self, name, email: str):
```
Hee, I did not know that! Still, I initially thought it was missing and you were unsure, so probably better to add `name: str` for clarity.

Will be addressed in https://github.com/timvink/mkdocs-git-authors-plugin/issues/8 so OK to leave for now

timvink commented on mkdocs_git_authors_plugin/repo.py:
```diff
@@ -0,0 +1,774 @@
+
+        Returns:
+            Unordered HTML list as a string.
+        """
```
My reasoning here was to separate the classes that interact with git with functionality that deals with the formatting. 

Template files sound interesting. Fits with the roadmap https://github.com/timvink/mkdocs-git-authors-plugin/issues/16 

I would be OK with creating a new issue for this to discuss, and leaving this as-is for now.

uliska replied:
ba91edd300df5e5775aa3e93d839929db52b734f moves this code to the util module as it is a formatting function. I will do so with the other formatting functions as well when I encounter them.

timvink replied:
Nice!

timvink commented on mkdocs_git_authors_plugin/repo.py:
```diff
@@ -0,0 +1,774 @@
+        """
+        return result
+
+    def commit(self, sha: str):
```
Yes, discussed earlier. resolving.

timvink commented on mkdocs_git_authors_plugin/repo.py:
```diff
@@ -0,0 +1,774 @@
+        cmd.run()
+        return cmd.stdout()[0]
+
+    def page(self, path):
```
yes, resolving.

timvink commented on mkdocs_git_authors_plugin/plugin.py:
```diff
@@ -1,15 +1,81 @@
     config_scheme = (
         ('show_contribution', config_options.Type(bool, default=False)),
+        ('show_lines', config_options.Type(bool, default=False)),
+        ('count_empty_lines', config_options.Type(bool, default=True)),
```
You're right of course.

timvink commented on mkdocs_git_authors_plugin/plugin.py:
```diff
@@ -1,15 +1,81 @@
+
+    def on_page_content(self, html, page, config, files, **kwargs):
+        """
+        Replace jinja tag {{ git_authors_list }} in HTML.
```
Let's discuss, because it's a big design decision for the plugin. My considerations:

Because I'm thinking about future integration with `git-revision-date`: 

1) I would like to keep it simple by having only one ``{{ git_authors_summary }}` tag. 
2) I would also like to keep the number of options down

Because I would like to promote integration into existing themes (most prominently `mkdocs-material`):

3) Have all the potentially relevant information exposed in `{{ git_authors }}` 
4) Promote users that want to do more than the simple summary to customize their used themes. (I updated the README to explain how to do this for `mkdocs-material` in 0668d3e51176a414e3a9410df4a79a26c31ab935)

So yes, being able to make an HTML list of authors is definitely in scope of the plugin, but we have not aligned yet on *how* do this properly.

uliska replied:
OK, I see. Please let this go through the PR and let us discuss this later, but have it clearly resolved in a way you're happy with before pushing a new release to PyPi. Right now I'm the only user of that functionality, and that will remain the case until it is available from PyPi. And I'm of course ready to adapt my few projects to whatever solution this will lead. 

I think the current state of it doesn't have serious side effects, it is "only" an additional function you don't know yet whether to keep. So I don't think there's much of a risk merging it temporarily.

uliska replied:
I've created #22 in order to mark this as resolved. Of course you can reopen it if you think it's necessary to add more discussion *here*.

timvink replied:
This is tough. I understand you have already implemented this PR in several projects.  

> The first rule of open source, according to `@shykes`: “No is temporary, yes is forever.”

I do not want to introduce extra markdown tags until we have a clear direction on how to deal with customization (I created https://github.com/timvink/mkdocs-git-authors-plugin/issues/21 for this), not even on unreleased master branch. 

I will merge this PR, and comment out `plugin.Plugin.on_page_content()`. You can fork, uncomment, and use that fork in your packages while we decide on #21.

timvink commented on mkdocs_git_authors_plugin/repo.py:
```diff
@@ -0,0 +1,774 @@
+    - stdout()
+    - stderr()
+    In case of an error a verbose GitCommandError is raised.
+    """
```
I just re-read the [Zen of Python](https://www.python.org/dev/peps/pep-0020/)

> There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.

I'm actually Dutch, but this point is not super obvious to me 😞😋. It's probably a matter of preference / coding style. The 'zen' does not mention the large file over one-file-per object.

For me it would be more readable to have a `git/` folder with smaller files for the main classes. But I won't press the issue, I'll leave it to you to decide.

uliska replied:
OK, maybe my recollection has misled me. Probably I've read some *comments* on the topic.

I agree that factoring out multiple modules may make the code cleaner, and maybe it will even help with one minor issue I had: I couldn't *always* provide type hinting because of circular dependencies between classes.

I agree to doing that, but only after a decision whether the Git classes will actually be kept.

#23

timvink commented:
> But it's important to clearly express differing opinions in order to be able to resolve them for the best.

I agree, and thanks for all your comments. I really appreciate the conversation ! 

Indeed some discussions about different programming styles, but educational for me as well. 

I resolved most discussions, couple of points where we need to think a bit more. Getting there!

timvink commented:
Merged! 🎉 Amazing work @uliska !
